### DISCUSSION TITLE
Best way to build with custom kernel

### DISCUSSION THREAD

**Discussion by liyimeng** [kairos-io/kairos#2240] [General]
I wanna build a variant with my own custom kernel. I make a possibilities exploration, as far as I found, there are some options:
- In Dockerfile.kairos-* add a custom repo, point to rebuild kernel packages.
- Create a luet kernel package in https://github.com/kairos-io/packages  reference it in kairos build

I am not sure what is the better way to go, any suggestion?

Thanks a lot in advanced!

**Comment by mauromorales** [comment]
@liyimeng if you use Ubuntu LTS, you can build your own image with the Kernel you want and feed it as BASE_IMAGE to the Dockerfile.kairos-ubuntu, then we will not install the kernel. For all other distros you will need to do that process yourself. IMO is better to do it in the Dockerfile because you still need to run dracut to build the initramfs at the end of the process, so it cannot be a static package comming for Luet only.

https://github.com/kairos-io/kairos/blob/master/images/Dockerfile.kairos-ubuntu#L166-L167

**Reply by mauromorales** [reply]
Here's a detailed example in case you want ubuntu non-hwe https://github.com/kairos-io/kairos/tree/master/examples/byoi/ubuntu-non-hwe but the process would be the same with any other kernel

**Reply by liyimeng** [reply]
@mauromorales Thanks a lot! I was not aware of dracut things. I'll give it a shot with your suggestion.

**Reply by mauromorales** [reply]
cool let me know if it works. I should have also pointed out to the docs, but I haven't tested the example in the docs recently so keep an eye open there. (the previous example I shared I tested last week so that should work)

https://kairos.io/docs/reference/build-from-scratch/

**Reply by liyimeng** [reply]
@mauromorales I have test it and confirm it works as you described.  However, I get nohang service keep spinning up and down, eating hard of my CPU resource.  I have 4G RAM. and free command report no memory issue. 

likely be another unknown issue. 

![Screenshot 2024-02-15 at 12 38 56](https://github.com/kairos-io/kairos/assets/3155688/1b3388b6-7236-4b6f-8aff-d773f3a28241)

**Reply by mauromorales** [reply]
yeah this might be related to the issue you mentioned on the other comment. When building the kernel, at the end of your Dockerfile, you might be able to disable nohang and see how that works for you. Maybe even completely remove the package

**Reply by mauromorales** [reply]
the method is still valid. You are building an image with the kernel or any other package that you need. Kernel is just a bit special in that you need to build the initrd afterwards but other than that is just building a new container image and then producing an iso (or any other artifact) with it

**Reply by liyimeng** [reply]
@mauromorales sorry, my fault, I look at the wrong branch, which contains old code

**Comment by liyimeng** [comment]
Likely be this one https://github.com/kairos-io/kairos/issues/2195
